### PR TITLE
[Connector-V2][JDBC-connector] support Jdbc dm

### PR DESCRIPTION
**Pull request by laglangyue**
## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

**Comment by laglangyue** [comment]
I imitate MySqlCatalog to DMCatalog, but I don't know it work.
and the workflows is failed, I need some time to solve it，and I try to learn know how to do it rencently.

**Comment by CalvinKirs** [comment]
If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

**Comment by CalvinKirs** [comment]
@ic4y PTAL,thx

**Comment by ic4y** [comment]
> I imitate MySqlCatalog to DMCatalog, but I don't know it work. and the workflows is failed, I need some time to solve it，and I try to learn know how to do it rencently.

Ask any questions at any time.

**Comment by laglangyue** [comment]
> If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

ok,I will lately

**Comment by laglangyue** [comment]
> > I imitate MySqlCatalog to DMCatalog, but I don't know it work. and the workflows is failed, I need some time to solve it，and I try to learn know how to do it rencently.
> 
> Ask any questions at any time.


TypeMapper is in the method 
org.apache.seatunnel.connectors.seatunnel.jdbc.catalog.MySqlCatalog#fromJdbcType

but also in this method
org.apache.seatunnel.connectors.seatunnel.jdbc.internal.dialect.mysql.MySqlTypeMapper

Any difference?

**Comment by laglangyue** [comment]
another question, Mysql has the catalog named MysqlCatalog ,and why does PostgreSQL have no catalog such as PostGresqlCatalog?

**Comment by ic4y** [comment]
> > > I imitate MySqlCatalog to DMCatalog, but I don't know it work. and the workflows is failed, I need some time to solve it，and I try to learn know how to do it rencently.
> > 
> > 
> > Ask any questions at any time.
> 
> TypeMapper is in the method org.apache.seatunnel.connectors.seatunnel.jdbc.catalog.MySqlCatalog#fromJdbcType
> 
> but also in this method org.apache.seatunnel.connectors.seatunnel.jdbc.internal.dialect.mysql.MySqlTypeMapper
> 
> Any difference?

There is really no difference between the two, This is indeed repeated in the implementation of the `catalog`. However, the catalog has not been started yet, so it is not necessary to implement the `catalog` related things.

Refer to the implementation under the `jdbc.internal.dialect` package.

**Comment by ic4y** [comment]
> another question, Mysql has the catalog named MysqlCatalog ,and why does PostgreSQL have no catalog such as PostGresqlCatalog?

Because only the catalog of mysql is currently implemented, pg has not yet been implemented. DM can also temporarily not implement catalog

**Comment by laglangyue** [comment]
> DM can also temporarily not implement catalog

thx，I get , I will modify my code this night

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/dm/DmdbTypeMapper.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    // ------------------------------blob-------------------------
+    public static final String DM_BLOB = "BLOB";
+    public static final String DM_BFILE = "BFILE";
+    public static final String DM_IMAGE = "IMAGE";
+    public static final String DM_LONGVARBINARY = "LONGVARBINARY";
+
+    @Override
+    public SeaTunnelDataType<?> mapping(ResultSetMetaData metadata, int colIndex) throws SQLException {
+        String dmdbType = metadata.getColumnTypeName(colIndex).toUpperCase();
+        //  String columnName = metadata.getColumnName(colIndex);
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/dm/DmdbTypeMapper.java]: Delete it if you don't need it

File: pom.xml [hunk 2]
```diff
@@ -352,6 +353,13 @@
                 <scope>test</scope>
             </dependency>
 
+            <dependency>
+                <groupId>com.dameng</groupId>
+                <artifactId>DmJdbcDriver18</artifactId>
+                <version>${dm-jdbc.version}</version>
+                <scope>provided</scope>
```
Comment (ic4y) [inline pom.xml]: this `scope` need to be `test`. We don't want to include DmJdbcDriver18.jar in the connector's package

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/dm/DmdbJdbcRowConverter.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+public class DmdbJdbcRowConverter extends AbstractJdbcRowConverter {
+
+    @Override
+    public String converterName() {
+        return "Dmdb";
+    }
+
+    @Override
+    public SeaTunnelRow toInternal(ResultSet rs, ResultSetMetaData metaData, SeaTunnelRowType typeInfo) throws SQLException {
+        return super.toInternal(rs, metaData, typeInfo);
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/dm/DmdbJdbcRowConverter.java:37]: Is it OK to use the default `toInternal`?, it is recommended to test the covered data types locally

**Comment by ic4y** [comment]
Have you tested XA-Transaction?, This will be used in exactly-once Sink. And add how to configure exactly-once  sink  in the document

**Comment by laglangyue** [comment]
> Have you tested XA-Transaction?, This will be used in exactly-once Sink. And add how to configure exactly-once sink in the document

Thx for your guide, I will do it this weekend.

**Comment by ic4y** [comment]
@laglangyue Wait for the JDBC e2e test (#2321) to merge. You need to add test for DM in jdbc e2e test

**Comment by laglangyue** [comment]
I have encountered a difficulty. The PostgreSQL driver is directly introduced and introduced into the Lib of connector-v2. I am afraid this is not appropriate.

In the E2E test, I need a driver in the docker. I can manually copy it  to the Lib folder of connector-v2-dist, but it will fail if in the CI environment or someone else.

Therefore, I wonder whether an additional module, similar to e2e-driver-jar-dist, is required to be provided to E2E test and copied to docker.

**Comment by hailin0** [comment]
> The PostgreSQL driver is directly introduced and introduced into the Lib of connector-v2. I am afraid this is not appropriate.

@CalvinKirs Do you have any suggestions

**Comment by laglangyue** [comment]
image
I have 2 ideas to solve it.
1. `maven-dependency-plugin` generate the dir such lib include all dirver-jar, and copy it to container, like above.
2. Run the command in container ` wget https://repo1.maven.org/maven2/com/dameng/DmJdbcDriver18/8.1.2.141/DmJdbcDriver18-8.1.2.141.jar`,and copy it to lib.

In our pom just need to provided scope
@CalvinKirs

**Comment by CalvinKirs** [comment]
> 

In fact, we could have a dist directory dedicated to these jars if we really needed to. (if necessary), but it seems that they are all jdbc driven, as a temporary solution, you can also use the second one. Of course, how you do it is up to you.

**Comment by laglangyue** [comment]
image

**Comment by laglangyue** [comment]
flink job is finish and there ara not any error log，but sink not has data

**Comment by laglangyue** [comment]
I has fix the date,time,timestamp convert in other PR，and the e2e test is finished.
the dirver of DM is Apache license in maven-center-reponsitory, so I add it. such as mysql/pg, how do you think it.
In my opinion, I will remove the driver from connector,and add dirver to e2e when the PR #2640 is merged.
Please review it angin @hailin0 @ic4y

**Comment by Hisoka-X** [comment]
Please fixed checkstyle problem and conflct

**Comment by laglangyue** [comment]
> Please fixed checkstyle problem and conflct

done

File: pom.xml [hunk 4]
```diff
@@ -220,7 +221,11 @@
                 <artifactId>postgresql</artifactId>
                 <version>${postgresql.version}</version>
             </dependency>
-
+            <dependency>
```
Comment (Hisoka-X) [inline pom.xml]: Hi, please move connector dependency to connector's pom.xml. Reference: #2630

**Comment by Hisoka-X** [comment]
@ic4y Hi, PTAL

File: pom.xml (hunk 4 above)
Comment (laglangyue) [reply pom.xml]: > Hi, please move connector dependency to connector's pom.xml. Reference: #2630

done

**Comment by laglangyue** [comment]
@ic4y @Hisoka-X @CalvinKirs  I has rebased it to dev, and change dependencies,Please review again.

File: tools/dependencies/known-dependencies.txt [hunk 5]
```diff
@@ -736,4 +736,5 @@ zstd-jni-1.3.3-1.jar
 zstd-jni-1.4.3-1.jar
 jakarta.activation-api-1.2.1.jar
 jakarta.xml.bind-api-2.3.2.jar
-postgresql-42.3.3.jar
\ No newline at end of file
+postgresql-42.3.3.jar
+DmJdbcDriver18-8.1.2.141.jar
```
Comment (ic4y) [inline tools/dependencies/known-dependencies.txt]: Now the connector and e2e do not need to add a license, delete it.

File: seatunnel-dist/release-docs/licenses/LICENSE-DMjdbc.txt [hunk 6]
```diff
@@ -0,0 +1,202 @@
+
```
Comment (ic4y) [inline seatunnel-dist/release-docs/licenses/LICENSE-DMjdbc.txt]: And do not need this file

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        assertHasData(SOURCE_TABLE);
+    }
+
+    @Test
+    @DisplayName("flink JDBC-DM test")
+    public void testJdbcDmdbSourceAndSink() throws IOException, InterruptedException, SQLException {
+        assertHasData(SOURCE_TABLE);
+        Container.ExecResult execResult = executeSeaTunnelFlinkJob("/jdbc/jdbc_dm_source_and_sink.conf");
+        Assertions.assertEquals(0, execResult.getExitCode());
+        assertHasData(SINK_TABLE);
```
Comment (ic4y) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java:145]: Suggest check whether the data in the two tables are the same

File: seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    public void testDMDBImage() {
+        assertHasData(SOURCE_TABLE);
+    }
+
+    @Test
+    @DisplayName("spark JDBC-DM test for all type mapper")
+    public void testDMDBSourceToJdbcSink() throws SQLException, IOException, InterruptedException {
+        Container.ExecResult execResult = executeSeaTunnelSparkJob("/jdbc/jdbc_dm_source_and_sink.conf");
+        Assertions.assertEquals(0, execResult.getExitCode());
+        assertHasData(SINK_TABLE);
```
Comment (ic4y) [inline seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java]: as above

File: seatunnel-examples/seatunnel-spark-connector-v2-example/pom.xml [hunk 9]
```diff
@@ -55,6 +55,11 @@
             <artifactId>connector-clickhouse</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.seatunnel</groupId>
+            <artifactId>connector-jdbc</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-spark-connector-v2-example/pom.xml]: revert?

File: seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml [hunk 10]
```diff
@@ -76,6 +76,11 @@
             <artifactId>connector-dingtalk</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.seatunnel</groupId>
+            <artifactId>connector-jdbc</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml]: revert?

File: seatunnel-e2e/seatunnel-spark-connector-v2-e2e/pom.xml [hunk 11]
```diff
@@ -55,6 +55,7 @@
                 </exclusion>
             </exclusions>
         </dependency>
+
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-spark-connector-v2-e2e/pom.xml:58]: revert?

File: seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        } catch (SQLException e) {
+            throw new RuntimeException("test dm server image error", e);
+        }
+    }
+
+    @Test
+    @DisplayName("JDBC-DM container can be pull")
+    public void testDMDBImage() {
+        assertHasData(SOURCE_TABLE);
+    }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java]: remove? 
 Only need to check sink data?

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    private static final String USERNAME = "SYSDBA";
+    private static final String PASSWORD = "SYSDBA";
+    private static final String DATABASE = "SYSDBA";
+    private static final String SOURCE_TABLE = "e2e_table_source";
+    private static final String SINK_TABLE = "e2e_table_sink";
+    private Connection jdbcConnection;
+    private GenericContainer<?> server;
+
+    @BeforeEach
+    public void startGreenplumContainer() throws ClassNotFoundException, SQLException {
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java]: Greenplum?

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+        server = new GenericContainer<>(DOCKER_IMAGE)
+            .withNetwork(NETWORK)
+            .withNetworkAliases(HOST)
+            .withLogConsumer(new Slf4jLogConsumer(log));
+        server.setPortBindings(Lists.newArrayList(
+            String.format("%s:%s", 5236, 5236)));
+        Startables.deepStart(Stream.of(server)).join();
+        log.info("Greenplum container started");
+        // wait for Greenplum fully start
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java]: Greenplum?

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+            String sql = String.format("select * from %s.%s limit 1", DATABASE, table);
+            ResultSet source = statement.executeQuery(sql);
+            Assertions.assertTrue(source.next());
+        } catch (SQLException e) {
+            throw new RuntimeException("test dm server image error", e);
+        }
+    }
+
+    @AfterEach
+    public void closeGreenplumContainer() throws SQLException {
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java]: Greenplum?

File: seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+                .untilAsserted(this::initializeJdbcConnection);
+            initializeJdbcTable();
+        } catch (Exception ex) {
+            log.error("dm container init failed", ex);
+            throw new RuntimeException(ex);
+        }
+    }
+
+    @AfterEach
+    public void closeGreenplumContainer() throws SQLException {
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java]: Greenplum?

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+        if (jdbcConnection != null) {
+            jdbcConnection.close();
+        }
+    }
+
+    @Test
+    @DisplayName("JDBC-DM container can be pull")
+    public void testDMDBImage() {
+        assertHasData(SOURCE_TABLE);
+    }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java:137]: remove?
Only need to check sink data?

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+        } catch (SQLException e) {
+            throw new RuntimeException("test dm server image error", e);
+        }
+    }
+
+    @AfterEach
+    public void closeGreenplumContainer() throws SQLException {
+        if (jdbcConnection != null) {
+            jdbcConnection.close();
+        }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java:127]: add close server

```java
if (server != null) {
    server.stop();
}
```

File: seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+        if (jdbcConnection != null) {
+            jdbcConnection.close();
+        }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java]: add close server

```java
if (dmServer != null) {
    dmServer.stop();
}
```

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java (hunk 7 above)
Comment (laglangyue) [reply seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java:145]: I Ihave debugged to check. Verifying the data consistency between sink and source is a public Issue.. I hope to discuss with you in another PR, such as how to preset data and assert verification. I think this is required for all E2E use cases. 
Currently, only DM has preset SQL for all fields, and other E2E has only a few fields

File: seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java (hunk 12 above)
Comment (laglangyue) [reply seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-jdbc-spark-e2e/src/test/java/org/apache/seatunnel/e2e/spark/v2/jdbc/JDBCDmdbIT.java]: this case is test the DDL/DML for Dmdb is ok. if someone modify the DDL/DML ,this testcase can work

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java (hunk 13 above)
Comment (laglangyue) [reply seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-jdbc-flink-e2e/src/test/java/org/apache/seatunnel/e2e/flink/v2/jdbc/JdbcDmdbIT.java]: done

**Comment by ic4y** [comment]
LGTM

**Review by Hisoka-X — APPROVED** [review]

**Review by EricJoy2048 — APPROVED** [review]
+1

**Pull request merged**